### PR TITLE
Bugfix FXIOS-10155 redux bug

### PR DESCRIPTION
**Pull request by OrlaM**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10155)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22258)

## :bulb: Description
The recent refactor to how redux actions execute surfaced a race condition where the webview was overwriting and removing the home page. This was previously happening in the opposite order so it wasn't causuing any obvious issues. I added a check to only update the webview when it is not the home page. 
I also cleaned up some unneded code as it was adding a lot of noise to the logs and making it difficult to debug. I think this code became unneded as more of BVC's logic is being routed through redux. 

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift [hunk 1]
```diff
@@ -160,35 +158,6 @@ struct BrowserViewControllerState: ScreenState, Equatable {
             microsurveyState: MicrosurveyPromptState.reducer(state.microsurveyState, action))
     }
 
-    static func reduceStateForPrivateModeAction(action: PrivateModeAction,
-                                                state: BrowserViewControllerState) -> BrowserViewControllerState {
-        switch action.actionType {
-        case PrivateModeActionType.privateModeUpdated:
```
Comment (OrlaM) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift:166]: An update is now triggered every time the a tab is selected so BVC no longer needs to listen for this action.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 2]
```diff
@@ -625,42 +625,40 @@ class BrowserViewController: UIViewController,
     }
 
     func newState(state: BrowserViewControllerState) {
-        ensureMainThread { [weak self] in
```
Comment (OrlaM) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:628]: Redux already checks for this so need to double up in the newState function.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 3]
```diff
@@ -3683,7 +3682,8 @@ extension BrowserViewController: TabManagerDelegate {
 
         scrollController.tab = selectedTab
 
-        if let webView = selectedTab.webView {
+        if !selectedTab.isFxHomeTab,
```
Comment (OrlaM) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:3685]: This is probably the only line of code actually needed.

File: firefox-ios/Client/Frontend/Toolbar+URLBar/OverlayModeManager.swift [hunk 4]
```diff
@@ -67,6 +67,7 @@ class DefaultOverlayModeManager: OverlayModeManager {
     }
 
     func finishEditing(shouldCancelLoading: Bool) {
+        guard inOverlayMode else { return }
```
Comment (OrlaM) [inline firefox-ios/Client/Frontend/Toolbar+URLBar/OverlayModeManager.swift:70]: This triggers an update of the entire BVC and adds a lot of noise to the logs so I added a check to only do this when needed.

**Comment by OrlaM** [comment]
@MergifyIO backport release/v132

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:3683]: nice!! 

> The recent refactor to how redux actions execute surfaced a race condition where the webview was overwriting and removing the home page. This was previously happening in the opposite order so it wasn't causuing any obvious issues.

Do you mind expanding more on this and why the opposite order wasn't causing any obvious issues? The new bug seems like a consistent and reproducible, so curious on how the redux surface this race condition.

**Review by cyndichin — APPROVED** [review]
testing on iPad / iPhone, looks great so approving! Thank you for finding this fix! 🔥 
I added a comment to better understand how you discovered the issue.

**Comment by cyndichin** [comment]
these PR seems to fail `testAutofillCreditCardsToggleOnOff` for all 3 times, not sure if these are flaky or not but seems highly unrelated

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
Comment (OrlaM) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:3683]: Previously the redux action logic was dispatching async to the main thread more often, no it executes immediately more often. Before it was calling this block first, which sets the webview content and removes the homepage view controller from the embedded view and then the redux actions related to the homepage change were being handled, putting the new homepage in place. After the refactor the home page action is being handled immediately, putting the home page in place, and then it gets removed by this block of code updating the webview. 
Not sure if I described that well as it's a bit messy 😓

**Comment by dragosb01** [comment]
> these PR seems to fail `testAutofillCreditCardsToggleOnOff` for all 3 times, not sure if these are flaky or not but seems highly unrelated

@OrlaM testAutofillCreditCardsToggleOnOff is passing on main, but on your branch it seems the card number has the query element modified. Not sure what part of your fix did that, but i have pushed a commit with the fix for this test.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:3683]: this helps! thank you :)

**Pull request merged**